pkg/cvo/sync_worker.go

```diff
@@ -200,12 +200,143 @@ func (w *SyncWorker) StatusCh() <-chan SyncWorkerStatus {
+		payloadUpdate, err := payload.LoadUpdate(info.Directory, desired.Image, w.exclude, w.includeTechPreview, w.clusterProfile)
+		if err != nil {
+			w.eventRecorder.Eventf(cvoObjectRef, corev1.EventTypeWarning, "VerifyPayloadFailed", "verifying payload failed version=%q image=%q failure=%v", desired.Version, desired.Image, err)
+			reporter.Report(SyncWorkerStatus{
+				Generation:  work.Generation,
+				Failure:     err,
+				Step:        "VerifyPayload",
+				Initial:     work.State.Initializing(),
+				Reconciling: work.State.Reconciling(),
+				Actual:      desired,
+				Verified:    info.Verified,
+			})
+			return err
+		}
+
+		payloadUpdate.VerifiedImage = info.Verified
+		payloadUpdate.LoadedAt = time.Now()
+
+		if work.Desired.Version == "" {
+			work.Desired.Version = payloadUpdate.Release.Version
+			desired.Version = payloadUpdate.Release.Version
+		} else if payloadUpdate.Release.Version != work.Desired.Version {
+			err = fmt.Errorf("release image version %s does not match the expected upstream version %s", payloadUpdate.Release.Version, work.Desired.Version)
+			w.eventRecorder.Eventf(cvoObjectRef, corev1.EventTypeWarning, "VerifyPayloadVersionFailed", "verifying payload failed version=%q image=%q failure=%v", work.Desired.Version, work.Desired.Image, err)
+			reporter.Report(SyncWorkerStatus{
+				Generation:  work.Generation,
+				Failure:     err,
+				Step:        "VerifyPayloadVersion",
+				Initial:     work.State.Initializing(),
+				Reconciling: work.State.Reconciling(),
+				Actual:      desired,
+				Verified:    info.Verified,
+			})
+			return err
+		}
+
+		// need to make sure the payload is only set when the preconditions have been successful
+		if len(w.preconditions) == 0 {
+			klog.V(4).Info("No preconditions configured.")
+		} else if info.Local {
+			klog.V(4).Info("Skipping preconditions for a local operator image payload.")
+		} else {
+			reporter.Report(SyncWorkerStatus{
+				Generation:  work.Generation,
+				Step:        "PreconditionChecks",
+				Initial:     work.State.Initializing(),
+				Reconciling: work.State.Reconciling(),
+				Actual:      desired,
+				Verified:    info.Verified,
+			})
+			if err := precondition.Summarize(w.preconditions.RunAll(ctx, precondition.ReleaseContext{DesiredVersion: payloadUpdate.Release.Version}, clusterVersion)); err != nil {
+				if work.Desired.Force {
+					klog.V(4).Infof("Forcing past precondition failures: %s", err)
+					w.eventRecorder.Eventf(cvoObjectRef, corev1.EventTypeWarning, "PreconditionsForced", "preconditions forced for payload loaded version=%q image=%q failures=%v", desired.Version, desired.Image, err)
+				} else {
+					w.eventRecorder.Eventf(cvoObjectRef, corev1.EventTypeWarning, "PreconditionsFailed", "preconditions failed for payload loaded version=%q image=%q failures=%v", desired.Version, desired.Image, err)
+					reporter.Report(SyncWorkerStatus{
```

Comment:
I'm a bit concerned about how the status consumer will handle two parallel streams of `Report` callers, now that we can have the payload-application loop and this update-acceptance loop going in parallel.  Maybe we need a new property in `SyncWorkerStatus` and then separate `status.condition` types for reporting?  I could see a `DesiredUpdateAccepted` or similar property that is unset when `desiredUpdate` is unset or matches our current target, goes `Unknown` when someone requests a new target and we're forming an opinion, goes `False` if we don't like the target, and goes away if we do like the target.  Or something like that.  I guess we could switch on `Step` without adding a new `SyncWorkerStatus` property, but something like `StepGroup` or similar might be more robust.